solidity/solidity-interfaces.md

```diff
@@ -0,0 +1,84 @@
+
+Interfaces are also valuable because they create a separation between source contracts and other
+contracts that need to interact with those source contracts. This separation significantly
+decreases compilation time when changes need to be made to source contracts. Any changes to a
+source contract would otherwise trigger a re-compilation of many other contracts that rely on that
+source contract, resulting in untolerably long waits.
+
+## Problem Statement
+
+We recently undertook an update to the Optimism Monorepo that added interfaces for all smart
+contracts. This resulted in a significant improvement to compilation time but simultaneously added
+complexity for any developer that needs to introduce a new smart contract. Most of this complexity
+arises because existing interface policy is non-obvious and the process for generating interfaces
+is entirely manual. We would like to make certain changes to this policy to support a better
+development process.
+
+We wanted to entirely automate the interface generation process so that users do not need to
+generate interfaces themselves. In exploratory work on this interface generation process, we've
+found that there are a number of important edge cases that came up.
+
+Specifically:
+
+- Source contracts using interfaces creates a basic dependency problem. For instance, suppose that
+  you'd like to add a new contract `ContractA`. You'd also like to create a new contract
+  `ContractB` that depends on `ContractA`. Our current policy states that `ContractB` should use
+  the interface for `ContractA`, `IContractA`. You would therefore need to generate the interface
+  for `ContractA` *before* starting to work on `ContractB`, which is annoying.
+- Circular dependencies in source contracts are basically impossible to resolve without manually
+  crafting contract interfaces. For instance, if `ContractA` depends on `IContractB` and
+  `ContractB` depends on `IContractA` but neither contract interface exists yet (new contracts)
+  then the contracts cannot be compiled (because neither interface exists) and the interfaces can
+  not be generated. Catch 22.
+- Using interfaces within the source contracts leads to other awkward situations with function
+  types that are challenging to resolve in an automated generation script.
+- Contract interfaces don't have source code, so navigating the functions in the source files
+  becomes awkward as each external call goes through an interface.
+
+## Proposed Solution
+
+All of the above seem to point to a change to our policy where contract interfaces are not used
+*anywhere* in the source contracts and are only used in tests or other external/peripheral
+locations that interact with the source contracts. Removing interfaces from source contracts
+entirely would also be a much clearer mental model ("just don't use interfaces in src/").
+
+In addition, interfaces would live in a new `interfaces` folder at the root of the contracts
```

Comment:
really like this, since my biggest issue with interface generation right now is I feel it generates way too many files and clutters up `src/`, and it makes it unclear when you are depending on machine generated interfaces vs [beautifully handwritten interfaces ](https://github.com/ethereum-optimism/optimism/blob/02d5832349a1f469adb00191459333e2032b09b4/packages/contracts-bedrock/src/dispute/interfaces/IBigStepper.sol#L7)